Feature/double precision

ethanwharris opened the pull request:
## What does this PR do?

Fixes #2497 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- For CHANGELOG separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

justusschock commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,52 @@
+    ) -> Tuple['Module', Sequence['Optimizer'], Sequence[Any]]:
+        """Converts the model to double precision and wraps the forward to convert incoming floating point data.
+        Does not alter `optimizers` or `lr_schedulers`."""
+        def patch_forward(old_forward: Callable) -> Callable:
```
can you move this as a static function to the class? So that it can be pickled?

justusschock commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,52 @@
+        """Converts the model to double precision and wraps the forward to convert incoming floating point data.
+        Does not alter `optimizers` or `lr_schedulers`."""
+        def patch_forward(old_forward: Callable) -> Callable:
+            def new_forward(*args, **kwargs) -> Any:
```
```suggestion
            @wraps(old_forward)
            def new_forward(*args, **kwargs) -> Any:
```

justusschock commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,52 @@
+                                   **apply_to_collection(kwargs, torch.Tensor, _to_double_precision))
+            return new_forward
+        model = model.to(dtype=torch.float64)
+        model.forward = patch_forward(model.forward)
```
So the model's forward will be patched forever? Can we maybe revert this once we finished training/testing/predicting?

justusschock commented on tests/plugins/test_double_plugin.py:
```diff
@@ -0,0 +1,42 @@
+class DoublePrecisionBoringModel(BoringModel):
+
+    def training_step(self, batch, batch_idx):
+        output = self((batch, torch.ones_like(batch).long()))  # Add some non floating-point data
```
can you also assert that here the data is of fp32? Just to make sure that we really convert and not our tests just pass because we passed already converted data :)

justusschock reviewed (commented):
Hi @ethanwharris ,

Great to see you again :)

And great PR!
Just some smaller comments :)

ethanwharris commented:
Hi @justusschock,

Thanks for the great suggestions, all done :)

ethanwharris commented:
The teardown is called at the end of training, but won't be called before checkpointing. Any way to address that?

williamFalcon commented:
omg.... i love this feature 
😍😍😍😍😍

rohitgr7 commented:
can we add support with deepspeed too?
https://github.com/PyTorchLightning/pytorch-lightning/blob/3a56a6024e0d8b239801cce558381807c24ba3d0/pytorch_lightning/plugins/training_type/deepspeed.py#L46-L48

not sure why it's handled this way in here, not an expert with deepspeed.

rohitgr7 commented on pytorch_lightning/trainer/connectors/accelerator_connector.py:
```diff
@@ -319,7 +320,8 @@ def select_precision_plugin(self) -> PrecisionPlugin:
             return PrecisionPlugin()
-
+        elif self.precision == 64:
+            return DoublePrecisionPlugin()
```
also update the error on line 363

ethanwharris replied:
Good catch! Done :)

carmocca commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,63 @@
+
+        @wraps(old_forward)
+        def new_forward(*args, **kwargs) -> Any:
+            return old_forward(*apply_to_collection(args, torch.Tensor, DoublePrecisionPlugin._to_double_precision),
+                               **apply_to_collection(kwargs, torch.Tensor, DoublePrecisionPlugin._to_double_precision))
```
You can apply the same pattern as here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/3a56a6024e0d8b239801cce558381807c24ba3d0/pytorch_lightning/plugins/training_type/deepspeed.py#L52-L58

By checking for `torch.DoubleTensor`

ethanwharris replied:
The method I used allowed for filtering using `torch.is_floating_point`, which is more reliable than just type checking for `FloatTensor` and `cuda.FloatTensor`. I can change this if needed but I think `is_floating_point` is a better approach. I modified this from the source for Nvidia Apex, which seems quite reliable :)

SeanNaren commented:
Thanks for the awesome contribution @ethanwharris!

Out of curiosity, if the user doesn't use the forward function of the LightningModule and instead calls independent modules, does it still work? i.e if I define my `training_step` like:

```python
 def training_step(self, batch, batch_idx):
        output = self.layer((batch, torch.ones_like(batch).long())) # use layer instead of self(...)
```

tchaton reviewed (approved):
Nice !

ethanwharris commented:
@SeanNaren That's an excellent point haha.

@SeanNaren @carmocca @justusschock I've made some changes to address your comments:
- Moved patch logic to seperate class that now handles teardown (much cleaner than before)
- Added patch to training_step etc. methods so that it all still works if the model is called directly without forward (and added a simple test of this behaviour)

@rohitgr7 Not sure about deepspeed, don't feel comfortable trying to add it as I don't currently have a way of testing it out. It seems that at least DeepSpeed ZeRO requires `precision=16`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/853523ee643fe0f0cc30d40d9e85a8869e7edfd8/tests/plugins/test_deepspeed_plugin.py#L193

Should all be ready for review now :smiley:

carmocca commented on tests/plugins/test_double_plugin.py:
```diff
@@ -0,0 +1,117 @@
+    trainer = Trainer(
+        max_epochs=2,
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
+        limit_test_batches=2,
+        limit_val_batches=2,
```
```suggestion
        fast_dev_run=2,
```

carmocca commented on tests/plugins/test_double_plugin.py:
```diff
@@ -0,0 +1,117 @@
+
+
+@pytest.mark.parametrize(
+    ['boring_model'],
+    [(DoublePrecisionBoringModel,), (DoublePrecisionBoringModelNoForward,)]
```
```suggestion
    'boring_model',
    (DoublePrecisionBoringModel, DoublePrecisionBoringModelNoForward)
```

carmocca commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+        self.method_name = method_name
+        self.old_method = old_method
+
+    def teardown(self):
```
```suggestion
    def teardown(self) -> None:
```

carmocca commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+class _DoublePrecisionPatch:
+    """Class to handle patching of methods in the `LightningModule` and subsequent teardown.`"""
+
+    def __init__(self, model: 'Module', method_name: str, old_method: Any):
```
```suggestion
    def __init__(self, model: 'Module', method_name: str, old_method: Any) -> None:
```

carmocca commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+        )
+
+    @staticmethod
+    def patch(model: 'Module', method_name: str):
```
```suggestion
    def patch(model: 'Module', method_name: str) -> _DoublePrecisionPatch:
```

carmocca commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+
+    precision: int = 64
+
+    def __init__(self):
```
```suggestion
    def __init__(self) -> None:
```

ethanwharris commented on tests/helpers/boring_model.py:
```diff
@@ -47,6 +47,20 @@ def __len__(self):
         return self.len
 
 
+class RandomFloatIntDataset(Dataset):
```
No problem, all done :)

rohitgr7 commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+                **_DoublePrecisionPatch._move_float_tensors_to_double(kwargs)
+            )
+
+        setattr(model, method_name, new_method)
```
```suggestion
        setattr(model, method_name, new_method if callable(old_method) else old_method)
```
just to avoid cases if some method is set to None, although not a good practice.

rohitgr7 commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+
+
+class DoublePrecisionPlugin(PrecisionPlugin):
+    """Plugin for training with double (`torch.float64`) precision."""
```
```suggestion
    """Plugin for training with double (``torch.float64``) precision."""
```

rohitgr7 commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+
+
+class _DoublePrecisionPatch:
+    """Class to handle patching of methods in the `LightningModule` and subsequent teardown.`"""
```
```suggestion
    """Class to handle patching of methods in the ``LightningModule`` and subsequent teardown."""
```

carmocca reviewed (approved):
LGTM. Can you have a look at the failing tests?

justusschock commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+            collection, torch.Tensor, function=_DoublePrecisionPatch._to_double_precision
+        )
+
+    @staticmethod
+    def patch(model: 'Module', method_name: str) -> '_DoublePrecisionPatch':
```
```suggestion
    @classmethod
    def patch(cls, model: 'Module', method_name: str) -> '_DoublePrecisionPatch':
```

justusschock commented on pytorch_lightning/plugins/precision/double.py:
```diff
@@ -0,0 +1,95 @@
+            )
+
+        setattr(model, method_name, new_method if callable(old_method) else old_method)
+        return _DoublePrecisionPatch(model, method_name, old_method)
```
```suggestion
        return cls(model, method_name, old_method)
```

justusschock reviewed (approved):
2 more minor comments, which allow subclassing (and don't hardcode the class name, else LGTM

ethanwharris commented:
@carmocca I have finally fixed the tests. `trainer.predict` calls `torch.set_grad_enabled(False)` which was then killing gradients for subsequent tests. I've added a call to `torch.set_grad_enable(True)` which fixes this for now. I'll open an issue to change the predict behaviour to re-enable grads, maybe just switch to `with torch.no_grad`.
